File: haskell/src/SwiftNav/SBP/Logging.hs
{-# OPTIONS_GHC -fno-warn-unused-imports #-}
{-# LANGUAGE NoImplicitPrelude           #-}
{-# LANGUAGE TemplateHaskell             #-}
{-# LANGUAGE RecordWildCards             #-}

-- |
-- Module:      SwiftNav.SBP.Logging
-- Copyright:   Copyright (C) 2015-2018 Swift Navigation, Inc.
-- License:     LGPL-3
-- Maintainer:  Swift Navigation <dev@swiftnav.com>
-- Stability:   experimental
-- Portability: portable
--
-- \<Logging and debugging messages from the device. \>

module SwiftNav.SBP.Logging
  ( module SwiftNav.SBP.Logging
  ) where

import BasicPrelude
import Control.Lens
import Control.Monad.Loops
import Data.Binary
import Data.Binary.Get
import Data.Binary.IEEE754
import Data.Binary.Put
import Data.ByteString.Lazy    hiding (ByteString)
import Data.Int
import Data.Word
import SwiftNav.SBP.TH
import SwiftNav.SBP.Types

{-# ANN module ("HLint: ignore Use camelCase"::String) #-}
{-# ANN module ("HLint: ignore Redundant do"::String) #-}
{-# ANN module ("HLint: ignore Use newtype instead of data"::String) #-}


msgLog :: Word16
msgLog = 0x0401

-- | SBP class for message MSG_LOG (0x0401).
--
-- This message contains a human-readable payload string from the device
-- containing errors, warnings and informational messages at ERROR, WARNING,
-- DEBUG, INFO logging levels.
data MsgLog = MsgLog
  { _msgLog_level :: !Word8
    -- ^ Logging level
  , _msgLog_text :: !Text
    -- ^ Human-readable string
  } deriving ( Show, Read, Eq )

instance Binary MsgLog where
  get = do
    _msgLog_level <- getWord8
    _msgLog_text <- decodeUtf8 . toStrict <$> getRemainingLazyByteString
    pure MsgLog {..}

  put MsgLog {..} = do
    putWord8 _msgLog_level
    putByteString $ encodeUtf8 _msgLog_text

$(makeSBP 'msgLog ''MsgLog)
$(makeJSON "_msgLog_" ''MsgLog)
$(makeLenses ''MsgLog)

msgFwd :: Word16
msgFwd = 0x0402

-- | SBP class for message MSG_FWD (0x0402).
--
-- This message provides the ability to forward messages over SBP.  This may
-- take the form of wrapping up SBP messages received by Piksi for logging
-- purposes or wrapping  another protocol with SBP.  The source identifier
-- indicates from what interface a forwarded stream derived. The protocol
-- identifier identifies what the expected protocol the forwarded msg contains.
-- Protocol 0 represents SBP and the remaining values are implementation
-- defined.
data MsgFwd = MsgFwd
  { _msgFwd_source    :: !Word8
    -- ^ source identifier
  , _msgFwd_protocol  :: !Word8
    -- ^ protocol identifier
  , _msgFwd_fwd_payload :: !Text
    -- ^ variable length wrapped binary message
  } deriving ( Show, Read, Eq )

instance Binary MsgFwd where
  get = do
    _msgFwd_source <- getWord8
    _msgFwd_protocol <- getWord8
    _msgFwd_fwd_payload <- decodeUtf8 . toStrict <$> getRemainingLazyByteString
    pure MsgFwd {..}

  put MsgFwd {..} = do
    putWord8 _msgFwd_source
    putWord8 _msgFwd_protocol
    putByteString $ encodeUtf8 _msgFwd_fwd_payload

$(makeSBP 'msgFwd ''MsgFwd)
$(makeJSON "_msgFwd_" ''MsgFwd)
$(makeLenses ''MsgFwd)

msgPrintDep :: Word16
msgPrintDep = 0x0010

-- | SBP class for message MSG_PRINT_DEP (0x0010).
--
-- Deprecated.
data MsgPrintDep = MsgPrintDep
  { _msgPrintDep_text :: !Text
    -- ^ Human-readable string
  } deriving ( Show, Read, Eq )

instance Binary MsgPrintDep where
  get = do
    _msgPrintDep_text <- decodeUtf8 . toStrict <$> getRemainingLazyByteString
    pure MsgPrintDep {..}

  put MsgPrintDep {..} = do
    putByteString $ encodeUtf8 _msgPrintDep_text

$(makeSBP 'msgPrintDep ''MsgPrintDep)
$(makeJSON "_msgPrintDep_" ''MsgPrintDep)
$(makeLenses ''MsgPrintDep)

File: kaitai/nim/logging.nim
import kaitai_struct_nim_runtime
import options

type
  Logging* = ref object of KaitaiStruct
    `parent`*: KaitaiStruct
  Logging_MsgLog* = ref object of KaitaiStruct
    `level`*: uint8
    `text`*: string
    `parent`*: Sbp_Message
  Logging_MsgFwd* = ref object of KaitaiStruct
    `source`*: uint8
    `protocol`*: uint8
    `fwdPayload`*: seq[uint8]
    `parent`*: Sbp_Message
  Logging_MsgPrintDep* = ref object of KaitaiStruct
    `text`*: string
    `parent`*: Sbp_Message

proc read*(_: typedesc[Logging], io: KaitaiStream, root: KaitaiStruct, parent: KaitaiStruct): Logging
proc read*(_: typedesc[Logging_MsgLog], io: KaitaiStream, root: KaitaiStruct, parent: Sbp_Message): Logging_MsgLog
proc read*(_: typedesc[Logging_MsgFwd], io: KaitaiStream, root: KaitaiStruct, parent: Sbp_Message): Logging_MsgFwd
proc read*(_: typedesc[Logging_MsgPrintDep], io: KaitaiStream, root: KaitaiStruct, parent: Sbp_Message): Logging_MsgPrintDep


proc read*(_: typedesc[Logging], io: KaitaiStream, root: KaitaiStruct, parent: KaitaiStruct): Logging =
  template this: untyped = result
  this = new(Logging)
  let root = if root == nil: cast[Logging](this) else: cast[Logging](root)
  this.io = io
  this.root = root
  this.parent = parent


proc fromFile*(_: typedesc[Logging], filename: string): Logging =
  Logging.read(newKaitaiFileStream(filename), nil, nil)


##[
This message contains a human-readable payload string from the device
containing errors, warnings and informational messages at ERROR,
WARNING, DEBUG, INFO logging levels.

]##
proc read*(_: typedesc[Logging_MsgLog], io: KaitaiStream, root: KaitaiStruct, parent: Sbp_Message): Logging_MsgLog =
  template this: untyped = result
  this = new(Logging_MsgLog)
  let root = if root == nil: cast[Logging](this) else: cast[Logging](root)
  this.io = io
  this.root = root
  this.parent = parent


  ##[
  Logging level

  ]##
  let levelExpr = this.io.readU1()
  this.level = levelExpr

  ##[
  Human-readable string

  ]##
  let textExpr = encode(this.io.readBytesFull(), "ascii")
  this.text = textExpr

proc fromFile*(_: typedesc[Logging_MsgLog], filename: string): Logging_MsgLog =
  Logging_MsgLog.read(newKaitaiFileStream(filename), nil, nil)


##[
This message provides the ability to forward messages over SBP.  This
may take the form of wrapping up SBP messages received by Piksi for
logging purposes or wrapping another protocol with SBP.

The source identifier indicates from what interface a forwarded stream
derived. The protocol identifier identifies what the expected protocol
the forwarded msg contains. Protocol 0 represents SBP and the remaining
values are implementation defined.

]##
proc read*(_: typedesc[Logging_MsgFwd], io: KaitaiStream, root: KaitaiStruct, parent: Sbp_Message): Logging_MsgFwd =
  template this: untyped = result
  this = new(Logging_MsgFwd)
  let root = if root == nil: cast[Logging](this) else: cast[Logging](root)
  this.io = io
  this.root = root
  this.parent = parent


  ##[
  source identifier

  ]##
  let sourceExpr = this.io.readU1()
  this.source = sourceExpr

  ##[
  protocol identifier

  ]##
  let protocolExpr = this.io.readU1()
  this.protocol = protocolExpr

  ##[
  variable length wrapped binary message

  ]##
  block:
    var i: int
    while not this.io.isEof:
      let it = this.io.readU1()
      this.fwdPayload.add(it)
      inc i

proc fromFile*(_: typedesc[Logging_MsgFwd], filename: string): Logging_MsgFwd =
  Logging_MsgFwd.read(newKaitaiFileStream(filename), nil, nil)


##[
Deprecated.

]##
proc read*(_: typedesc[Logging_MsgPrintDep], io: KaitaiStream, root: KaitaiStruct, parent: Sbp_Message): Logging_MsgPrintDep =
  template this: untyped = result
  this = new(Logging_MsgPrintDep)
  let root = if root == nil: cast[Logging](this) else: cast[Logging](root)
  this.io = io
  this.root = root
  this.parent = parent


  ##[
  Human-readable string

  ]##
  let textExpr = encode(this.io.readBytesFull(), "ascii")
  this.text = textExpr

proc fromFile*(_: typedesc[Logging_MsgPrintDep], filename: string): Logging_MsgPrintDep =
  Logging_MsgPrintDep.read(newKaitaiFileStream(filename), nil, nil)

